sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NativeMethodsShouldBeWrapped.cs

```diff
@@ -0,0 +1,131 @@
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class NativeMethodsShouldBeWrapped : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4200";
+        private const string MessageFormat = "{0}";
+        private const string MakeThisMethodPrivateMessage = "Make this native method private and provide a wrapper.";
+        private const string MakeThisWrapperLessTrivialMessage = "Make this wrapper for native method '{0}' less trivial.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(ReportPublicExternalMethods, SyntaxKind.MethodDeclaration);
+            context.RegisterSyntaxNodeAction(ReportTrivialWrappers, SyntaxKind.MethodDeclaration);
+        }
+
+        private static void ReportPublicExternalMethods(SyntaxNodeAnalysisContext c)
+        {
+            var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+
+            if (methodDeclaration.Modifiers.Any(IsExternal) &&
+                methodDeclaration.Modifiers.Any(IsPublicOrProtected))
+            {
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, methodDeclaration.Identifier.GetLocation(),
+                    MakeThisMethodPrivateMessage));
+            }
+        }
+
+        private static void ReportTrivialWrappers(SyntaxNodeAnalysisContext c)
+        {
+            var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+            var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+
+            if (methodDeclaration.Modifiers.Any(IsExternal) ||
+                methodDeclaration.ParameterList?.Parameters.Count == 0 ||
+                HasTwoOrMoreStatements(methodDeclaration))
+            {
+                return;
+            }
+
+            var invocations = GetBodyDescendants(methodDeclaration)
+                .OfType<InvocationExpressionSyntax>();
+
+            if (invocations.Take(2).Count() == 2) // Two or more invocations
+            {
+                return;
+            }
+
+            bool ParametersMatchCotnainingMethodDeclaration(InvocationExpressionSyntax invocation) =>
+                invocation.ArgumentList.Arguments
+                    .All(a => a.Expression is LiteralExpressionSyntax ||

```

Comment:
The expression inside of Any is a bit complicated. Perhaps put them in separate method?